src/cluster.c

```diff
@@ -852,23 +692,231 @@ static dict *parse_cluster_slots(valkeyClusterContext *cc, valkeyReply *reply) {
     return NULL;
 }
 
+/* Keep lists of parsed replica nodes in a dict using the primary_id as key. */
+static int retain_replica_node(dict *replicas, char *primary_id, valkeyClusterNode *node) {
+    sds key = sdsnew(primary_id);
+    if (key == NULL)
+        return VALKEY_ERR;
+
+    struct hilist *replicaList;
+
+    dictEntry *de = dictFind(replicas, key);
+    if (de == NULL) {
+        /* Create list to hold replicas for a primary. */
+        replicaList = listCreate();
+        if (replicaList == NULL) {
+            sdsfree(key);
+            return VALKEY_ERR;
+        }
+        replicaList->free = listClusterNodeDestructor;
+        if (dictAdd(replicas, key, replicaList) != DICT_OK) {
+            sdsfree(key);
+            listRelease(replicaList);
+            return VALKEY_ERR;
+        }
+    } else {
+        sdsfree(key);
+        replicaList = dictGetEntryVal(de);
+    }
+
+    if (listAddNodeTail(replicaList, node) == NULL)
+        return VALKEY_ERR;
+
+    return VALKEY_OK;
+}
+
+/* Store parsed replica nodes in the primary nodes, which holds a list of replica
+ * nodes. The `replicas` dict shall contain lists of nodes with primary_id as key. */
+static int store_replica_nodes(dict *nodes, dict *replicas) {
+    if (replicas == NULL)
+        return VALKEY_OK;
+
+    dictIterator di;
+    dictInitIterator(&di, nodes);
+    dictEntry *de;
+    while ((de = dictNext(&di))) {
+        valkeyClusterNode *primary = dictGetEntryVal(de);
+
+        /* Move replica nodes related to this primary. */
+        dictEntry *der = dictFind(replicas, primary->name);
+        if (der != NULL) {
+            assert(primary->slaves == NULL);
+            /* Move replica list from replicas dict to nodes dict. */
+            primary->slaves = dictGetEntryVal(der);
+            dictSetHashVal(replicas, der, NULL);
+        }
+    }
+    return VALKEY_OK;
+}
+
+/* Parse a node from a single CLUSTER NODES line. Returns an allocated
+ * valkeyClusterNode as a pointer in `parsed_node`.
+ * Only parse primary nodes if the `parsed_primary_id` argument is NULL,
+ * otherwise replicas are also parsed and its primary_id is returned by pointer
+ * via 'parsed_primary_id'. */
+static int parse_cluster_nodes_line(valkeyClusterContext *cc, char *line,
+                                    valkeyClusterNode **parsed_node, char **parsed_primary_id) {
+    char *p, *id = NULL, *addr = NULL, *flags = NULL, *primary_id = NULL,
+             *link_state = NULL, *slots = NULL;
+    /* Find required fields and keep a pointer to each field:
+     * <id> <addr> <flags> <primary_id> <ping-sent> <pong-recv> <config-epoch> <link-state> [<slot> ...]
+     */
+    // clang-format off
+    int i = 0;
+    while ((p = strchr(line, ' ')) != NULL) {
+        *p = '\0';
+        switch(i++){
```

Comment:
Clang-format didn't mention this?

```suggestion
        switch (i++){
```

Reply:
It's disabled here to get a compact switch, but then this was missed.. :smile:
I now also see that there should be an additional space before `{`..